cmd/soroban-cli/src/commands/config/data.rs

```diff
@@ -0,0 +1,203 @@
+    };
+    let id = ulid::Ulid::new();
+    let file = actions_dir()?.join(id.to_string()).with_extension("json");
+    std::fs::write(file, serde_json::to_string(&data)?)?;
+    Ok(id)
+}
+
+pub fn read(id: &ulid::Ulid) -> Result<(Action, Uri), Error> {
+    let file = actions_dir()?.join(id.to_string()).with_extension("json");
+    let data: Data = serde_json::from_str(&std::fs::read_to_string(file)?)?;
+    Ok((data.action, http::Uri::from_str(&data.rpc_url)?))
+}
+
+pub fn write_spec(hash: &str, spec_entries: &[xdr::ScSpecEntry]) -> Result<(), Error> {
+    let file = spec_dir()?.join(hash);
+    tracing::trace!("writing spec to {:?}", file);
+    let mut contents: Vec<u8> = Vec::new();
+    for entry in spec_entries {
+        contents.extend(entry.to_xdr(xdr::Limits::none())?);
+    }
+    std::fs::write(file, contents)?;
+    Ok(())
+}
+
+pub fn read_spec(hash: &str) -> Result<Vec<xdr::ScSpecEntry>, Error> {
+    let file = spec_dir()?.join(hash);
+    tracing::trace!("reading spec from {:?}", file);
+    Ok(soroban_spec::read::parse_raw(&std::fs::read(file)?)?)
+}
+
+pub fn list_ulids() -> Result<Vec<ulid::Ulid>, Error> {
+    let dir = actions_dir()?;
+    let mut list = std::fs::read_dir(dir)?
+        .map(|entry| {
+            entry
+                .map(|e| e.file_name().into_string().unwrap())
+                .map_err(Error::from)
+        })
+        .collect::<Result<Vec<String>, Error>>()?;
+    list.sort();
+    Ok(list
+        .iter()
+        .map(|s| ulid::Ulid::from_str(s.trim_end_matches(".json")))
+        .collect::<Result<Vec<_>, _>>()?)
+}
+
+pub fn list_actions() -> Result<Vec<DatedAction>, Error> {
+    list_ulids()?
+        .into_iter()
+        .rev()
+        .map(|id| {
+            let (action, uri) = read(&id)?;
+            Ok(DatedAction(id, action, uri))
+        })
+        .collect::<Result<Vec<_>, Error>>()
+}
+
+pub struct DatedAction(ulid::Ulid, Action, Uri);
+
+impl std::fmt::Display for DatedAction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let (id, a, uri) = (&self.0, &self.1, &self.2);
+        let datetime = to_datatime(id).format("%b %d %H:%M");
+        let status = match a {
+            Action::Simulation(sim) => sim
+                .error
+                .as_ref()
+                .map_or_else(|| "SUCCESS".to_string(), |_| "ERROR".to_string()),
+            Action::Transaction(txn) => txn.status.to_string(),
+        };
+        write!(f, "{id} {} {status} {datetime} {uri} ", a.type_str(),)
+    }
+}
+
+impl DatedAction {}
+
+fn to_datatime(id: &ulid::Ulid) -> chrono::DateTime<chrono::Utc> {
+    chrono::DateTime::from_timestamp_millis(id.timestamp_ms().try_into().unwrap()).unwrap()
+}
+
+#[derive(Serialize, Deserialize)]
+struct Data {
+    action: Action,
+    rpc_url: String,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub enum Action {
+    Simulation(SimulateTransactionResponse),
+    Transaction(GetTransactionResponseRaw),
```

Comment:
I think we should call this Send instead of Transaction, assuming it's only recorded when sending a tx. The action names should correlate to commands that someone can run.